HPCC-35300 LDAP should handle brute force password hack attempts

kenrowland opened the pull request:
Added cache of failed user authentication attempts with configurable number 
of allowed failed attempts and lockout timeout

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Copilot committed:
HPCC-35300 LDAP should handle brute force password hack attempts

Added cache of failed user authentication attempts with configurable
number of allowed failed attempts and lockout timeout

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

jakesmith commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -27,6 +27,30 @@
+
+// Default configuration values
+static constexpr unsigned DEFAULT_MAX_FAILED_ATTEMPTS = 5;
+static constexpr unsigned DEFAULT_FAILED_AUTH_CACHE_TIMEOUT_SECONDS = 300; // 5 minutes
```
style: preferred as camel case - all caps preferred for macros only,

jakesmith commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -570,6 +594,67 @@ ISecProperty* CLdapSecResourceList::findProperty(const char* name)
 
+// Check if a user is in the failed authentication cache and should be blocked
+// Returns true if user should be blocked, false if authentication should proceed
+static bool checkFailedAuthCache(const char* username, unsigned maxFailedAttempts, unsigned cacheTimeoutSeconds)
```
why are these static methods, I think makes more sense if members of CLdapSecManager where they're used, and they have access to maxfailedattempts and cachetimeoutseconds.

jakesmith commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -694,6 +781,23 @@ bool CLdapSecManager::authenticate(ISecUser* user)
+    // Check failed authentication cache before proceeding
+    if (checkFailedAuthCache(username, m_maxFailedAuthAttempts, m_failedAuthCacheTimeout))
+    {
+        DBGLOG("User %s authentication blocked due to failed authentication cache", username);
```
I agree, *but* I'm nervous about this emitting too much logging. We really need a general helper function that only emits if not emited within T time.

jakesmith reviewed (changes requested):
@kenrowland - please see comments.

kenrowland commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -27,6 +27,30 @@
+
+// Failed authentication cache using username as key
+static std::map<std::string, FailedAuthEntry> g_failedAuthCache;
+static CriticalSection g_failedAuthCacheLock;
```
Added a max size limit. Then, when inserting or removing an entry, the cache is scanned for any additional items that can be removed based on cache timeout. This seems like a good way to limit the entries w/o doing unnecessary scans. Note that the size can grow past the limit if all existing entries have not exceeded their cache timeouts yet.

kenrowland commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -694,6 +781,23 @@ bool CLdapSecManager::authenticate(ISecUser* user)
+    // Check failed authentication cache before proceeding
+    if (checkFailedAuthCache(username, m_maxFailedAuthAttempts, m_failedAuthCacheTimeout))
+    {
+        DBGLOG("User %s authentication blocked due to failed authentication cache", username);
```
Agreed. I opened https://hpccsystems.atlassian.net/browse/HPCC-35681 to review all the logging in the LDAP security manager as a separate task.

Commit:
Address review comments
Convert cache to class to better encapsulate functionality
Small refactoring to authentication to centralize failed
auth cache management

jakesmith commented on system/security/LdapSecurity/failedAuthCache.hpp:
```diff
@@ -0,0 +1,160 @@
+        if (it == m_cache.end())
+        {
+            time_t currentTime = time(nullptr);
+            m_cache[username] = FailedAuthEntry(currentTime, 1);
```
similar comments isUserLockedOut, this code could be simplieid and use common code by using CTimedCache. e.g.
```
unsigned *attempts = cache.query(username, true);
if (nullptr == attempts)
    cache.add(username, 1);
else
{
    (*attempts)++;
    if (*attempts == m_maxFailedAttempts)
    {
    }
}
```

NB: the 'true' to query, will cause the time to be updated. That's not what the PR currently does, but I think it's what you want.
e.g. imagine with 5 min cache timeout:
00:01:00 - userX - isUserLockedOut (miss) - authenticate fails - entry in cache created   - failedAttempts = 1
00:03:00 - userX - isUserLockedOut (hit)    - authenticate fails - entry in cache updated - failedAttempts = 2
00:04:00 - userX - isUserLockedOut (hit)    - authenticate fails - entry in cache updated - failedAttempts = 3
00:04:59 - userX - isUserLockedOut (hit)    - authenticate fails - entry in cache updated - failedAttempts = 4
00:05:02 - userX - isUserLockedOut (miss) - authenticate fails - entry in cache created   - failedAttempts = 1

Because the time is created when the cache entry is added, and never updated, it times out after that time, even if the last failed auth has just passed.
If there's a flood of failed authentications, it shouldn't timeout, the fails should reset the clock.

CTimedCache::query(...., true) will do that.

jakesmith commented on system/security/LdapSecurity/ldapsecurity.cpp:
```diff
@@ -714,10 +646,19 @@ void CLdapSecManager::init(const char *serviceName, IPropertyTree* cfg)
 
+    // Initialize/update the shared failed-auth cache with configured values
+    {
+        CriticalBlock block(CLdapSecManager::s_failedAuthCacheInitLock);
```
not great encapsulation. This should be using the crit in the cache class..
To do that, I think you should have an init function in the FailedAuthCache class that holds the crit.

But it _should_ be immutable, it's only not because it's shared, and it seems flawed if >1 CLdapSecManager initialized these settings.

jakesmith commented on system/security/LdapSecurity/failedAuthCache.hpp:
```diff
@@ -0,0 +1,160 @@
+        {
+            time_t currentTime = time(nullptr);
+            m_cache[username] = FailedAuthEntry(currentTime, 1);
+            trimFailedAuthCache();
```
I think you should call only call CTimedCache::clearTimedoutEntries() here, and only if has hit the limit cache limit. This is the only place that can increase it's size.

jakesmith reviewed (changes requested):
@kenrowland - please see comments.

The shared FailedAuthCache instance, means multiple CLdapSecManagers would all configure the instance with potentially different settings, which makes no sense (and theoretically could lead to it going wrong).

If it's going to shared, it needs to have a single shared configuration, and be initialized/configured once.

kenrowland commented:
Going to re-implement using CTimedCache suggestion. CTimedCache is a private class in Dali utils that needs to be shared from JLib. https://hpccsystems.atlassian.net/browse/HPCC-35793 was opened to do this work. Once shared, a new PR will be created.
